### DISCUSSION TITLE
Large-scale code reformatting

### DISCUSSION THREAD

**Discussion by greglandrum** [rdkit/rdkit#7850] [Development]
As an FYI: I just merged a PR (#7849) containing formatting changes to a bunch of the .cpp and .h files. 
This is the result of running clang-format-18 across the whole code base so all of the C++ source should now be consistently formatted.

I know that this kind of large-scale change can be painful for people with open PRs, but we need to do it eventually and this can only be done immediately before a major release.

**Comment by mwojcikowski** [comment]
@greglandrum Should there be an action that does check the code formatting on both Python and Cpp on each PR to avoid such cases in the end? Also you can add a  .gitignore file and add merge hash to avoid polluting git blame.

**Reply by greglandrum** [reply]
> @greglandrum Should there be an action that does check the code formatting on both Python and Cpp on each PR to avoid such cases in the end? 

That sounds like a reasonable thing to do (not that I know how to do it), but I'm happy to hear other people's feedback on that. It does increase the friction in creating a PR, but maybe that is work doing?

Before anyone brings it up: I am not happy adding any kind of action which automatically reformats the code as part of the PR process. I'm fine having something that checks whether or not code *would* be re-formatted, but not actually doing the formatting.

> Also you can add a .gitignore file and add merge hash to avoid polluting git blame.

I will look into this

**Reply by mwojcikowski** [reply]
[Here](https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/) is some docs from github. Sorry for confusion with gitignore.

I can make a pr for automated actions with **checks** next week, unless someone beats me to it.

**Reply by bp-kelley** [reply]
Not running clang-format is something I am very guilty of before pushing a PR.  I think the right action is to add a pre-commit hook to the forked repos  (not the main one, as I don't believe to can run one on a pull request) and perhaps make this a recommendation for developers.  The setup is pretty straightforward:

https://pre-commit.com

**Reply by DavidACosgrove** [reply]
Is it not something that an IDE can do?  I use CLion on a Mac, and it has settings for “action on save” which includes running clang on the code.

**Reply by bp-kelley** [reply]
I've tried in vain to get Xcode (which I use sometimes begrudgingly because it has a great debugger) to do, but the pre-commit hook works for all my IDEs ...

**Reply by mwojcikowski** [reply]
Indeed formatting should be done by the commiter. The action would solely prohibit code that does not meet required criteria. I would throw black formatting on top of what Greg has already suggested. 

Precomitt  works too, although in my experience it require at least some sort of unified environment as the linters are non-standard third party libraries one has to install anyways.

**Reply by greglandrum** [reply]
> Indeed formatting should be done by the commiter. The action would solely prohibit code that does not meet required criteria. I would throw black formatting on top of what Greg has already suggested.

A minor thing: we currently use yapf for the Python code (not black) and unless there's a really compelling reason to switch I would prefer to keep it that way.
